src/pallene/gc.lua

```diff
@@ -28,77 +29,234 @@ local types = require "pallene.types"
---
---   4) Use SSA form or some form of reaching definitions analysis so that we we only need to mirror
+--   3) Use SSA form or some form of reaching definitions analysis so that we we only need to mirror
 --      the writes that reach a GC site, instead of always mirroring all writes to a variable if one
 --      of them reaches a GC site.
 
 local gc = {}
 
-function gc.compute_stack_slots(func)
+local function FlowState()
+    return {
+        input  = {},  -- {var_id -> bool?} live variables at block start
+        output = {},  -- {var_id -> bool?} live variables at block end
+        kill   = {},  -- {var_id -> bool?} variables that are killed inside block
+        gen    = {},  -- {var_id -> bool?} variables that become live inside block
+    }
+end
 
-    local flat_cmds = ir.flatten_cmd(func.blocks)
+function gc.cmd_uses_gc(tag)
+    assert(tagged_union.typename(tag) == "ir.Cmd")
+    return tag == "ir.Cmd.CallStatic" or
+           tag == "ir.Cmd.CallDyn" or
+           tag == "ir.Cmd.CheckGC"
+end
 
-    -- 1) Compute approximated live intervals for GC variables defined by the function. Function
-    -- parameters are only counted if they are redefined, since their original value was already
-    -- saved by the caller. Also note that we only care about variables, not about upvalues.
-    -- The latter are already exposed to the GC via the function closures.
+local function copy_set(S)
+    local new_set = {}
+    for v,_ in pairs(S) do
+        new_set[v] = true
+    end
+    return new_set
+end
 
-    local defined_variables = {} -- { var_id }, sorted by first definition
-    local last_use          = {} -- { var_id => integer }
-    local first_definition  = {} -- { var_id => integer }
+local function flow_analysis(block_list, state_list)
+    local function apply_gen_kill_sets(flow_state)
+        local input = flow_state.input
+        local output = flow_state.output
+        local gen = flow_state.gen
+        local kill = flow_state.kill
+        local in_changed = false
 
-    for i, cmd in ipairs(flat_cmds) do
-        for _, val in ipairs(ir.get_srcs(cmd)) do
-            if val._tag == "ir.Value.LocalVar" then
-                local v_id = val.id
-                last_use[v_id] = i
+        for v, _ in pairs(output) do
+            local val = true
+            if kill[v] then
+                val = nil
             end
+            local previous_val = input[v]
+            local new_val = previous_val or val
+            input[v] = new_val
+            in_changed = in_changed or (previous_val ~= new_val)
         end
-        for _, v_id in ipairs(ir.get_dsts(cmd)) do
-            local typ = func.vars[v_id].typ
-            if types.is_gc(typ) and not first_definition[v_id] then
-                first_definition[v_id] = i
-                table.insert(defined_variables, v_id)
+
+        for v, g in pairs(gen) do
+            assert(g ~= true or g ~= kill[v], "gen and kill can't both be true")
```

Comment:
Could it be `assert(not (g and kill[v]))`?
Checking against `==true` is too specific.

Reply:
Yeah, it's the same thing. I'll change it to `assert(not (g and kill[v]))`